Add documentation for Flannel

lastcoolnameleft opened the pull request:
/kind documentation

This PR is to add documentation to support Flannel in a CAPZ cluster.  No issue was created for it.


- [ ] squashed commits
- [X] includes documentation
- [ ] adds unit tests

**Release note**:
```release-note
Add documentation for using Flannel as CNI
```

shysank commented:
/ok-to-test

shysank commented:
@lastcoolnameleft Thanks for the pr! Can you create an issue for this for tracking? Also you'll have to update the [summary](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/docs/book/src/SUMMARY.md) page.

devigned reviewed (changes requested):
Please add a release note.

Thank you adding this documentation. At some point in the future, perhaps, we should consider turning this into a custom resource set.

lastcoolnameleft commented:
I have:

* Created Issue #1375 
* Updated the summary page
* Added a release note (to this issue.  Let me know if I need to somewhere else)

Anything else I should do?

shysank commented on docs/book/src/topics/flannel.md:
````diff
@@ -0,0 +1,61 @@
+Before deploying the cluster, change the `KubeadmControlPlane` value at `spec.kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs.allocate-node-cidrs` to `"true"`
+
+```yaml
+apiVersion: controlplane.cluster.x-k8s.io/v1alpha3
````
```suggestion
apiVersion: controlplane.cluster.x-k8s.io/v1alpha4
```

shysank commented on docs/book/src/topics/flannel.md:
````diff
@@ -0,0 +1,61 @@
+View your capi-cluster.yaml and make note of the Cluster Network CIDR Block.  For example: 
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha3
````
```suggestion
apiVersion: cluster.x-k8s.io/v1alpha4
```

shysank commented on docs/book/src/topics/flannel.md:
````diff
@@ -0,0 +1,61 @@
+
+```yaml
+kind: ConfigMap
+apiVersion: v1
````
```suggestion
apiVersion: v1
metadata:
  name: kube-flannel-cfg
```

shysank commented on docs/book/src/topics/flannel.md:
````diff
@@ -0,0 +1,61 @@
+        "Type": "vxlan"
+      }
+    }
+```
````
I understand it's kind of implied to apply `kube-flannel.yml` after this step, should we make it explicit here for completeness?

lastcoolnameleft replied:
Works for me.  Will add.

lastcoolnameleft replied:
Done.

jsturtevant commented:
>  At some point in the future, perhaps, we should consider turning this into a custom resource set.

We use flannel for windows and have a CRS for it https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/templates/addons/flannel-resource-set.yaml.  It currently refrences the flannel under a windows folder but the flannel deployment is valid for linux (the linux nodes are using it) https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/templates/addons/windows/flannel.yaml. This could be refactored to not include windows and roll it all up.

Something I would add to these docs is setting the MTU correctly for through put.  Some details in https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1353 but essentially vxlan on azure needs to be below the 1400 MTU of Azure's network

lastcoolnameleft commented:
> Something I would add to these docs is setting the MTU correctly for through put. Some details in #1353 but essentially vxlan on azure needs to be below the 1400 MTU of Azure's network

Looking at the issue setting MTU to 1392 appears to be a unvalidated value for this setting.  Should we wait until it's been validated?  [Set to 1392](https://github.com/flannel-io/flannel/blob/master/Documentation/configuration.md#key-command-line-options)?  Or proceed with the documentation as is?

lastcoolnameleft commented:
> Please add a release note.
> 
> Thank you adding this documentation. At some point in the future, perhaps, we should consider turning this into a custom resource set.

@devigned I added a release note to this PR.  Is that what you're asking for?

CecileRobertMichon commented:
@lastcoolnameleft thanks for adding the release note in the PR description. Can you also please squash commits before merge?

CecileRobertMichon commented on docs/book/src/topics/flannel.md:
```diff
@@ -0,0 +1,68 @@
+# Flannel
```
I wonder if there should be some mention of using Calico in here since that's what we use in tests and quickstart docs right now but there's no mention of it anywhere in this book so it might be confusing to users 🤔  What do you think about adding a short sentence about other CNIs in here and pointing to the Calico template and/or opening a follow up issue for adding more comprehensive docs on Azure + Calico?

lastcoolnameleft commented:
@CecileRobertMichon Commits squashed and I added a new sentence about installing your own CNI solution with a link to the Quickstart CNI walkthrough

CecileRobertMichon commented:
/lgtm

shysank commented:
/lgtm

devigned reviewed (approved):
/lgtm
/approve

Thank you!
